src/data/ProgramsD2Repository.ts

```diff
@@ -0,0 +1,185 @@
+import _ from "lodash";
+import { ClosePatientsOptions, ProgramsRepository } from "domain/repositories/ProgramsRepository";
+import { Async } from "domain/entities/Async";
+import { D2Api } from "types/d2-api";
+import { TrackedEntity } from "domain/entities/TrackedEntity";
+import log from "utils/log";
+
+export class ProgramsD2Repository implements ProgramsRepository {
+    constructor(private api: D2Api) {}
+
+    async closePatients(options: ClosePatientsOptions): Async<void> {
+        const {
+            orgUnitsIds,
+            startDate,
+            endDate,
+            programId,
+            programStagesIds,
+            closureProgramId,
+            timeOfReference,
+            pairsDeValue,
+            comments,
+            post,
+        } = options;
+
+        const daysOfReference = parseInt(timeOfReference);
+        if (_.isNaN(daysOfReference)) throw new Error("Time of reference must be a number");
+
+        const filteredEntities$ = this.api
+            .get<ApiGetResponse>("/tracker/trackedEntities", {
+                program: programId,
+                orgUnit: orgUnitsIds?.join(";"),
+                ouMode: orgUnitsIds ? "SELECTED" : "ALL",
+                enrollmentOccurredAfter: startDate,
+                enrollmentOccurredBefore: endDate,
+                fields: "*,enrollments[*]",
+                skipPaging: true,
+            })
+            .map(({ data }) => {
+                const teisWithoutClosure = data.instances.filter(tei => {
+                    const enrollment = _.first(tei.enrollments);
+                    if (!enrollment) return;
```

Comment:
[very minor] `filter` should always return a boolean (it's a predicate). It can be moved to the final condition: `enrollment && ...`